content/api/commands/origin.md

```diff
@@ -0,0 +1,482 @@
+
+Callbacks are executed inside an entirely separate instance of Cypress, so
+arguments must be transmitted to the other instance by means of
+[the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).
+The interface for this mechanism is the `args` option.
+
+```js
+const sentArgs = { username: 'username', password: 'P@55w0rd!' }
+cy.origin(
+  'supersecurelogons.com',
+  // Send the args here...
+  { args: sentArgs },
+  // ...and receive them at the other end here!
+  ({ username, password }) => {
+    cy.visit('/login')
+    cy.get('input#username').type(username)
+    cy.get('input#password').type(password)
+    cy.contains('button', 'Login').click()
+  }
+)
+```
+
+Note: You can just replace `sentArgs` and `receivedArgs` with `args` if you
+want, the naming in this example is purely for clarity.
+
+### Navigating to secondary origin with cy.visit
+
+When navigating to a secondary origin using `cy.visit()`, it is essential to
+trigger the navigation **after** entering the origin callback, otherwise a
+cross-domain error will be thrown.
+
+```js
+// Do things in primary domain...
+
+cy.origin('https://www.acme.com', () => {
+  // Visit https://www.acme.com/history/founder
+  cy.visit('/history/founder')
+  cy.get('h1').contains('About our Founder, Marvin Acme')
+})
+```
+
+TODO baseUrl / default protocol
+
+### Navigating to secondary origin with UI
+
+When navigating to a secondary origin by clicking a link or button in the
+primary origin, it is essential to trigger the navigation _before_ entering the
+origin callback, otherwise a cross-domain error will be thrown.
+
+```js
+// Button in primary origin goes to https://www.acme.com
+cy.contains('button', 'Go').click()
+
+cy.origin('www.acme.com', () => {
+  // No cy.visit is needed as the button brought us here
+  cy.get('h1').contains('ACME CORP')
+})
+```
+
+### Navigating to multiple secondary origins in succession
+
+Callbacks may **not** themselves contain `cy.origin()` calls, so when visiting
+multiple origins, do so at the top level of the test.
+
+```js
+cy.origin('foo.com', () => {
+  cy.visit('/')
+  cy.url().should('contain', 'foo.com')
+})
+
+cy.origin('bar.com', () => {
+  cy.visit('/')
+  cy.url().should('contain', 'bar.com')
+})
+
+cy.origin('baz.com', () => {
+  cy.visit('/')
+  cy.url().should('contain', 'baz.com')
+})
+```
+
+<Alert type="info">
+
+A future version of Cypress will allow the use of nested `cy.origin()` calls.
+
+</Alert>
+
+### SSO login custom command
+
+A very common requirement is logging in to a site before running a test. If
+login itself is not the specific focus of the test, it's good to encapsulate
+this functionality in a `login`
+[custom command](/api/cypress-api/custom-commands) so you don't have to
+duplicate this login code in every test. Here's an idealized example of how to
+do this with `cy.origin()`.
+
+```js
+Cypress.Commands.add('login', (username, password) => {
```

Comment:
I wonder if we should just demonstrate the example using `cy.session()`? I'm worried someone will see this example and copy it, not realizing they should use the below example.

Reply:
I like the approach of gradually increasing complexity, but I agree that blind copying is a possibility, so I added an "Inefficient Usage" warning & icon